### PR TITLE
Issue 200 expression tree notebook

### LINKED ISSUE #54: optimise adjacent scalars in tree after set parameters
After `ParameterValues` swaps out all the `Parameter` for `Scalar` nodes, there is the possibility that you get things like `Scalar(4) * Scalar(5)` in the tree, which can be combined into `Scalar(20)`

### LINKED ISSUE #200: expression tree structure notebook
see #199

### PR DESCRIPTION
**Pull request by martinjrobins**
# Description

- added notebook illustrating pybamm expression trees
- altered Symbol visualise function to write out to the given filename (i.e. don't add a directory or extension) @Scottmar93: see what you think of this
- simplified the StateVector name to improve visualisation of tree
- added `simplify` function to Symbol, which aims to provide basic simplifications to an expression tree. 
- simplified Matrix name to improve visualisation of the tree

Fixes #200 
Fixes #54

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Key checklist:

- [ ] No style issues: `$ flake8`
- [ ] All tests pass: `$ python run-tests.py --unit`
- [ ] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks: 

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: examples/notebooks/expression-tree.ipynb [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "D = pybamm.Parameter('D')\n",
+    "c = pybamm.Variable('c', domain=['negative electrode'])\n",
+    "\n",
+    "rhs_c = D * pybamm.div(pybamm.grad(c))\n",
```
Comment (tinosulzer) [inline examples/notebooks/expression-tree.ipynb]: It might be clearer to call this `dcdt` instead of `rhs_c`

Comment (martinjrobins) [reply examples/notebooks/expression-tree.ipynb]: agreed

File: pybamm/expression_tree/binary_operators.py [hunk 2]
```diff
@@ -108,6 +114,24 @@ def evaluate(self, t=None, y=None):
         """ See :meth:`pybamm.Symbol.evaluate()`. """
         return self.children[0].evaluate(t, y) + self.children[1].evaluate(t, y)
 
+    def simplify(self):
+        """ See :meth:`pybamm.Symbol.simplify()`. """
+
+        # helper function to see if node evaluates to zero
+        def is_zero(node):
+            return node.evaluates_to_number() and node.evaluate() == 0
```
[the remark this answers is not in the document]
Comment (martinjrobins) [reply pybamm/expression_tree/binary_operators.py]: yea, I think that would be useful. I've added `evaluates_to_value` in Symbol to make it more flexible

**Comment by martinjrobins** [comment]
> Is there any reason you've only implemented those simplifications? There would be many more possible ones

I just wanted to put in what I needed for this issue, I agree there is many more to add.

**Comment by tinosulzer** [comment]
Great, feel free to merge when this is ready. And let's try to keep the momentum going for documentation :)

**Comment by martinjrobins** [comment]
well, looks like I need to implement the other binary simplifications to ensure testing coverage of `BinaryOperator`, doing these now....

**Pull request merged**